### ISSUE TITLE
Error using indexed access types with mapped types

### ISSUE THREAD

**Issue by irudoy** [microsoft/TypeScript#49242]
# Bug Report

### 🔎 Search Terms

indexed access types, mapped types

### 🕗 Version & Regression Information

- This changed between versions 4.6.4 and 4.7.0

### ⏯ Playground Link

[v4.8.0-dev.20220525 ❌](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220525#code/C4TwDgpgBAKuEGcoF4oG8CwAoKUBmAlgE4LABc6UAhgIwXBECu0AvgNza4IQDGA9gDsAJhTTUATPSasOOKMAAWxEZSoBmKcyjtsLbNh4AbKgiQxEwdJygQBDAolFQA2jCgEBUANYQQfPLDwCAC6APwUcJAIrsHOwdr6cvwCpEw8wHxEABQAlFZyuIoECAB0tvaIKOg6cnpY1lRCQgCidkQgADxuEAAewLZCSD5+AZGIAHxZAlQAthARADQ2bSARQTF5mAXuAVkAhEWl5UQO0dNzwZvWuIVKR22nzucQ8ahxsjcJ9duHZQ+IT1mL1CJTAjAQCiyxxAOQ+XzqQA)

[v4.7.0-beta ❌](https://www.typescriptlang.org/play?ts=4.7.0-beta#code/C4TwDgpgBAKuEGcoF4oG8CwAoKUBmAlgE4LABc6UAhgIwXBECu0AvgNza4IQDGA9gDsAJhTTUATPSasOOKMAAWxEZSoBmKcyjtsLbNh4AbKgiQxEwdJygQBDAolFQA2jCgEBUANYQQfPLDwCAC6APwUcJAIrsHOwdr6cvwCpEw8wHxEABQAlFZyuIoECAB0tvaIKOg6cnpY1lRCQgCidkQgADxuEAAewLZCSD5+AZGIAHxZAlQAthARADQ2bSARQTF5mAXuAVkAhEWl5UQO0dNzwZvWuIVKR22nzucQ8ahxsjcJ9duHZQ+IT1mL1CJTAjAQCiyxxAOQ+XzqQA)

[v4.6.4 ✅](https://www.typescriptlang.org/play?ts=4.6.4#code/C4TwDgpgBAKuEGcoF4oG8CwAoKUBmAlgE4LABc6UAhgIwXBECu0AvgNza4IQDGA9gDsAJhTTUATPSasOOKMAAWxEZSoBmKcyjtsLbNh4AbKgiQxEwdJygQBDAolFQA2jCgEBUANYQQfPLDwCAC6APwUcJAIrsHOwdr6cvwCpEw8wHxEABQAlFZyuIoECAB0tvaIKOg6cnpY1lRCQgCidkQgADxuEAAewLZCSD5+AZGIAHxZAlQAthARADQ2bSARQTF5mAXuAVkAhEWl5UQO0dNzwZvWuIVKR22nzucQ8ahxsjcJ9duHZQ+IT1mL1CJTAjAQCiyxxAOQ+XzqQA)

[v4.5.5 ❌](https://www.typescriptlang.org/play?ts=4.5.5#code/C4TwDgpgBAKuEGcoF4oG8CwAoKUBmAlgE4LABc6UAhgIwXBECu0AvgNza4IQDGA9gDsAJhTTUATPSasOOKMAAWxEZSoBmKcyjtsLbNh4AbKgiQxEwdJygQBDAolFQA2jCgEBUANYQQfPLDwCAC6APwUcJAIrsHOwdr6cvwCpEw8wHxEABQAlFZyuIoECAB0tvaIKOg6cnpY1lRCQgCidkQgADxuEAAewLZCSD5+AZGIAHxZAlQAthARADQ2bSARQTF5mAXuAVkAhEWl5UQO0dNzwZvWuIVKR22nzucQ8ahxsjcJ9duHZQ+IT1mL1CJTAjAQCiyxxAOQ+XzqQA)

### 💻 Code

```ts
type Types = {
  first: { a1: true };
  second: { a2: true };
  third: { a3: true };
}

class Test {
  entries: { [T in keyof Types]?: Types[T][] }

  constructor() {
    this.entries = {};
  }

  addEntry<T extends keyof Types>(name: T, entry: Types[T]) {
    if (!this.entries[name]) {
      this.entries[name] = [];
    }

    this.entries[name]?.push(entry); // error
  }
}
```

### 🙁 Actual behavior

Got the following error:
```
Argument of type '{ a1: true; } | { a2: true; } | { a3: true; }' is not assignable to parameter of type '{ a1: true; } & { a2: true; } & { a3: true; }'.
  Type '{ a1: true; }' is not assignable to type '{ a1: true; } & { a2: true; } & { a3: true; }'.
    Property 'a2' is missing in type '{ a1: true; }' but required in type '{ a2: true; }'
```

### 🙂 Expected behavior

No error. Or maybe I missed something, and this was intentional.

Seems like it was fixed in https://github.com/microsoft/TypeScript/pull/47109 and broken in v4.7.

**Comment by jcalz** [comment]
Looks like something's going on when the mapped type is optional.  This happened between 4.7.0-dev.20220302 and 4.7.0-dev.20220329.  It would definitely be nice if this didn't stay broken in 4.7.x for long.

**Comment by ahejlsberg** [comment]
Yeah, this is a bug. We're being a bit too conservative in the `isMappedTypeGenericIndexedAccess` function. It disallows any `?` modifiers when really it should only disallow `-?`. Easy fix, I'll have a PR up soon.
